README.md
====================================
README for '3DS soil model' (soil3ds)
====================================

This is '3DS soil model' (soil3ds), a 3D Soil model of soil for water and N balances adapted from the STICS soil module.

See 
- Louarn G, Faverjon L, Migault V, Escobar-Gutiérrez A, Combes D. (2016). Assessment of ‘3DS’, a soil module for individual-based models of plant communities. In: IEEE International Conference on Functional-Structural Plant Growth Modeling, Simulation, Visualization and Applications (FSPMA), 125–132. doi: 10.1109/FSPMA.2016.7818298
- N. Brisson, M. Launay, B. Mary, N. Beaudoin (2008) Conceptual basis, formalisations and parameterization of the STICS crop model, Quae, Versailles.
- N. Brisson & A. Perrier (1991). A semi-empirical model of bare soil evaporation for crop simulation models. Water resources research, 27(5), 719-727. 
- Lebon, E., Dumas, V., Pieri, P., & Schultz, H. R. (2003). Modelling the seasonal dynamics of the soil water balance of vineyards. Functional Plant Biology, 30(6), 699-710.


## 1. Getting Started

These instructions will get you a copy of *soil3ds* up and running on your local 
machine.

### 1.1 Prerequisites

To install and use *soil3ds*, you need first to install the dependencies.

*soil3ds* has been tested on Windows.
 
#### 1.1.1 Install the dependencies on Windows 10 64 bit
1) Create a conda environment with miniconda3
    ```bash
    conda create -n myenvname python=3.7 xlrd=2.0.1 numpy=1.20.3 scipy=1.7.3 pandas=1.3.4
    ```

2) Place yourself in the created environment  : `conda activate myenvname`

3) Install *soil3ds*
    1) Git console :
        ```bash
        git clone -b Develop https://github.com/glouarn/soil3ds.git
        ```
    2) installation in the conda environment (in folder `soil3ds`)
        ```bash
        python setup.py develop
        ```


### 1.3 Running

To run a simulation example :

* 1. place yourself in folder `soil3ds/test`
  2. run from the console:
		```bash
        python test_soil_coupling1.py
        ```


## 2. Reading the docs

To build the user and reference guides:


## 3. Testing

The test allows to verify that the model implementation accurately 
represents the developer’s conceptual description of the model and its solution.


## Contact

For any question, send an email to <gaetan.louarn@inrae.fr>.

## Authors

**Gaëtan LOUARN**, **Eric LEBON** - see file [AUTHORS](AUTHORS) for details

## License

This project is licensed under the CeCILL-C License - see file [LICENSE](LICENSE) for details


CHANGES
soil3ds 1.1
--------------------

- version compatible with python 3 (3.7)


soil3ds 1.0
--------------------

First version of 3DS-soil-model.


soil3ds/test/test_unit.py

import os
import sys
import soil3ds
path_ = os.path.dirname(os.path.abspath(soil3ds.__file__))  # local absolute path of L-egume
print(('path', path_))
sys.path.insert(0, path_)

from soil3ds.soil_moduleW import * #soil3ds installe comme module

################## TESTS ###############################

def test_uni1():
    """
    
    """
    ###############
    ## DEMO 1
    ## sol 1D (S)
    ## 1 seul systeme racinaire en dynamique
    ###############

    ## meteo
    Et0 = 4.
    precips = [0.]*29+[90.]# precipitations = 20mm tous les 30 jours
    precips = precips*20
    Precip = precips[0]#0.
    Irrig = 0.

    ## sol
    par_sol = {'13':{'KST': '5', 'teta_ad': 0.0078506705623363031, 'soil number': '13', 'teta_fc': 0.35816688969184857, 'WCST': '0.503', 'teta_wp': 0.11250451238171076, 'soil type': '13 - loam', 'teta_sat': 0.503, 'gamma_theo': '0.08489778'}}

    dz_sol = 5. #cm
    ncouches_sol = 40
    soil_type = 13 #"13 - loam"
    ZESX = 0.3#m
    CFES = 1.
    Uval = 3.#U quantite d'eau dans une couche superieure en mm (5 par default)
    b = 0.63

    ##vegetation
    espi_t = [0.312710721209,0.430217175269,0.527633447259,0.608394373323,0.675347532642,0.745670599554,0.827513705872,0.884631623932,0.923898514961,0.950492197719,0.968236501678,0.979901799595,0.987458191053,0.992281421006,0.995315225515,0.312710721209,0.430217175269,0.527633447259,0.608394373323,0.675347532642,0.745670599554,0.827513705872,0.884631623932,0.923898514961,0.950492197719,0.968236501678,0.979901799595,0.987458191053,0.992281421006,0.995315225515,0.312710721209,0.430217175269,0.527633447259,0.608394373323,0.675347532642,0.745670599554,0.827513705872,0.884631623932,0.923898514961,0.950492197719,0.968236501678,0.979901799595,0.987458191053,0.992281421006,0.995315225515,0.312710721209,0.430217175269,0.527633447259,0.608394373323,0.675347532642,0.745670599554,0.827513705872,0.884631623932,0.923898514961,0.950492197719,0.968236501678,0.979901799595,0.987458191053,0.992281421006]
    ls_epsi = [espi_t[0]]

    ## soil initialisation
    S = Soil(par_sol, soil_number = [soil_type]*ncouches_sol, dxyz = [[1.], [1.], [dz_sol/100.]*ncouches_sol], vDA=[1.25]*ncouches_sol, ZESX=ZESX, CFES=CFES)

    #RL_profil = table[0] #profil exporte de l-py via table
    RL_profil = RLprof_t(10, ncouches_sol)
    R1 = vert_roots(S.dxyz, RL_profil)#*100. /  (S.m_soil_vol*100.*100*100.)#profil densite racinaire en cm.cm-3
    ls_roots = [R1]#[R1, R2]#[R3]#

    ##bouble journaliere
    ls_transp, evapo_tot, D, state,  m_frac_transpi, m_frac_evap, ls_ftsw = S.stepWBmc(Et0, ls_roots, ls_epsi, Precip, Irrig,previous_state=[0.,0.,0.], ZESX=ZESX, leafAlbedo=0.15, U=Uval, b=b, FTSWThreshold=0.4, treshEffRoots=0.5, opt=1)
    for i in range(1,len(espi_t)):
        previous_asw = sum3(S.asw_t)
        Precip = 0.#precips[i]
        Irrig = 0.
        ls_epsi = [espi_t[i]]

        RL_profil = RLprof_t(10+i, ncouches_sol)#table[i]
        R1 = vert_roots(S.dxyz, RL_profil)#*100. /  (S.m_soil_vol*100.*100*100.)#profil densite racinaire en cm.cm-3
        ls_roots = [R1]#[R1, R2]#[R3]#

        ls_transp, evapo_tot, D, state,  m_frac_transpi, m_frac_evap, ls_ftsw =  S.stepWBmc(Et0, ls_roots, ls_epsi, Precip, Irrig, state, ZESX=ZESX, leafAlbedo=0.15, U=Uval, b=0.63, FTSWThreshold=0.4, treshEffRoots=0.5, opt=1)

        print(i, ls_ftsw, sum3(S.asw_t),  sum3(m_frac_evap), sum3(m_frac_transpi[0]), state)

    #    ##visu
    #    #bx = Box(Vector3(1.,1.,1.))
    #    Monviewer=Viewer
    #    MaScene=Scene()
    #    MaScene = S.plot_soil_properties (vals=S.ftsw_t, MaScene=Scene(), col_scale=5)#plot_soilWC (S.dxyz, S.m_soil_vox, S.ftsw_t, MaScene)
    #    Monviewer.display(MaScene)



def test_uni2():
    """
    
    """
    #################
    ## DEMO 2
    ## avec sol 3D (S2)
    ## 2 systemes racinaires fixes en competition (homogenes horizontalement) R3/R4
    ## en statique
    #################


    ##meteo
    Et0 = 2.
    #precips = [0.]*29+[90.]# precipitations = 20mm tous les 30 jours
    #precips = precips*20
    precips = [0.]*500#pas de precipitations (dessechement)
    Precip = precips[0]#0.
    Irrig=0.

    ## sol
    par_sol = {'13':{'KST': '5', 'teta_ad': 0.0078506705623363031, 'soil number': '13', 'teta_fc': 0.35816688969184857, 'WCST': '0.503', 'teta_wp': 0.11250451238171076, 'soil type': '13 - loam', 'teta_sat': 0.503, 'gamma_theo': '0.08489778'}}

    dz_sol = 5. #cm
    ncouches_sol = 10
    soil_type = 13 #"13 - loam"
    ZESX = 0.3 #m
    CFES=1.
    b=0.63
    Uval = 3.#sum(S.asw_t[0])#U quantite d'eau dans une couche superieure en mm (5 par default)
    stateEV = [0.,0.,0.]

    ### soil initialisation
    S2 = Soil(par_sol, soil_number = [soil_type]*ncouches_sol, dxyz = [[0.2]*5, [0.2]*5, [0.2]*10], vDA=[1.25]*ncouches_sol, ZESX=ZESX, CFES=CFES)

    ## vegetation
    ls_epsi = [0.4, 0.4]##[0.4]# !! correspondance avec les nb de root systems!
    ## 3D root distributions (fixe)
    R3 = zeros(shape(S2.m_1))
    R3[0:5, 1:3, 1:3] = ones([5,2,2])*0.25
    R4 = zeros(shape(S2.m_1))
    R4[0:7, 0:4, 2:4] = ones([7,4,2])*0.15

    ls_roots = [R3, R4]

    ##bouble journaliere
    ls_transp, evapo_tot, D, stateEV,  m_frac_transpi, m_frac_evap, ls_ftsw = S2.stepWBmc(Et0, ls_roots, ls_epsi, Precip, Irrig, stateEV, ZESX=ZESX, leafAlbedo=0.15, U=Uval, b=b, FTSWThreshold=0.4, treshEffRoots=0.5, opt=1)
    S2.update_ftsw()

    for i in range(40):
        Precip = precips[i]
        ls_transp, evapo_tot, D, stateEV,  m_frac_transpi, m_frac_evap, ls_ftsw = S2.stepWBmc(Et0, ls_roots, ls_epsi, Precip, Irrig, stateEV, ZESX=ZESX, leafAlbedo=0.15, U=Uval, b=b, FTSWThreshold=0.4, treshEffRoots=0.5, opt=1)
        S2.update_ftsw()

    #    ##visu
    #    Monviewer=Viewer
    #    MaScene=Scene()
    #    MaScene = S2.plot_soil_properties (vals=S2.ftsw_t, MaScene=Scene(), col_scale=5)
    #    Monviewer.display(MaScene)

    print((S2.ftsw_t))

print("test 1")
test_uni1()

#print("test 2")
#test_uni2()


soil3ds/miscel_functions.py
'''

 3DS-soil-model, a 3D Soil model adapted from STICS : miscellaneous soil functions
 ******************************
 Authors: G. Louarn 
 

 

'''


from scipy import *
import numpy as np
from copy import deepcopy


########## fonctions diverses matrice sol 3D
def sum3(mat):
    """
    
    """
    return mat.sum()#sum(sum(sum(mat)))
    ## pour gerer corectement les sommes totale un peu plus compliquee en muldim...


def sum_mat(ls_m):
    """
    
    """
    #pour gerer correctement les somme de matrices similaires dans une liste (avant sum le faisait bien, mais derniere version bug???)
    s = ls_m[0]
    if len(ls_m)>1:
        for i in range(1,len(ls_m)):
            s=s+ls_m[i]

    return s


def mask(mat, tresh=0.):
    """
    
    """
    #""" retourne matrice masque de 0  et 1
    #utile pour root/no roots; asw/no asw """
    #m = deepcopy(mat) #pas utile->mat cree nouvelle matrice
    return where(mat > tresh, 1, 0)
    #m = m*1. #pour convertir en float

#mask(R2)
#mask(asw_t)


########## diverses fonction parametrage sol

def tetavol_pF_curve(par_sol, pF_):
    """
    
    """
    #""" empirical relationship between volumic water content and pF (=ln(abs(h))where h is soil moisture succion )
    #Driessen, 1986 cited by Penning de Cries et al., 1989
    #WCST = volumic water content at saturation
    #gamma = empirical parameter to account for bulk density and texture"""
    
    teta = float(par_sol['WCST'])*exp(-float(par_sol['gamma_theo'])*pF_*pF_)
    return teta
    #tetavol_pF_curve(par_sol, 4.2)


def default_tetaref(par_sol, fc=2., wp=4.2, ad=7.):
    """
    
    """
    par_sol['teta_sat'] = tetavol_pF_curve(par_sol, 0.)
    par_sol['teta_fc'] = tetavol_pF_curve(par_sol, fc)
    par_sol['teta_wp'] = tetavol_pF_curve(par_sol, wp)
    par_sol['teta_ad'] = tetavol_pF_curve(par_sol, ad)
    return par_sol
    #p = default_tetaref(par_sol)


def pF(h):
    """
    
    """
    return log10(abs(h))




########## diverses fonction - soil water balance



## soil evaporation

def bEV(ACLIMc, ARGIs, HXs):
    """
    
    """
    #""" calcul de  b: coeff empirique sans dimension pour l'evaporation du sol a partir de (STICS manual p 128) : 
    #    ACLIMc : wheather parameter qui depend de la vitesse moyenne du vent (1m-1-> 20; 2m.s-1-> 14)
    #    ARGIs : clay content de la couche superficielle (%)
    #    HXs: volumetric moisture content at fc of the surface layer """
    
    HA = ARGIs/1500.
    return 0.5*ACLIMc*(HXs-HA)*power(0.63-HA, 5./3.)

#bEV(ACLIMc=20, ARGIs=20, HXs=0.4)



def soil_EV_1C(Et0, Precip, epsi, previous_state = [0.,0.,0.], leafAlbedo=0.15, U=5., b=0.63):
    """
    
    """
    #""" U: reservoir superficiel = quantite d'eau dans une couche superieure (varie generallement entre 0 et 30 mm
    #    b: coeef empirique sans dimension pour l'evaporation du sol (cf fonction bEV)
    #    meme formulation pour Lebon et al 2003 et STICS"""

    #recup previous state
    previousSES1 = previous_state[0]
    previousSES2 = previous_state[1]
    previousSEP = previous_state[2]

    coeffTV = epsi/ (1.-leafAlbedo)
    potentialSoilEvaporation = (1. - coeffTV) * Et0
    EP = (1. - coeffTV) * Et0
        
    SES1=max(0., previousSES1)
    SES2=max(0.,previousSES2)
    SEP=max(0.,previousSEP)
    
    P = Precip
    P1 = Precip 
    
    if (SES1<U):
        # phase 1
        if (P1>SES1):
            SES1 = 0.
            SES1 = SES1 + EP
        else:
            SES1 = SES1 - P
            SES1 = SES1 + EP
        
        if (SES1>U):
            ER = EP - 0.4*(SES1-U)
            SES2 = 0.6*(SES1-U)
        else:
            ER = EP
    else:
        # phase 2
        if (P1>SES2):
            SEP = 0
            P1 = P1 - SES2
            SES1 = U - P1
            
            if (P1>U):
                SES1 = EP
            else:
                SES1 = SES1 + EP
            
            if (SES1>U):
                ER = EP - 0.4*(SES1-U)
                SES2 = 0.6*(SES1-U)
            else:
                ER = EP
            
        else:
            SES21 = ((2*b*SEP + b**2)**0.5) - b
            SEP = SEP + EP
            SES20 = ((2*b*SEP + b**2)**0.5) - b
            ER = SES20 - SES21
            
            if (P>0):
                ESx = 0.8*P
                
                if (ESx<ER):
                    ESx = ER + P
                
                if (ESx>EP):
                    ESx = EP
                
                ER = ESx
                SES2 = SES2 + ER - P
            else:
                if (ER>EP):
                    ER = EP
                    SES2 = SES2 + ER - P
                else:
                    SES2 = SES2 + ER - P

    state = [SES1, SES2, SEP]
    soilEvaporation = ER 
    return soilEvaporation, state
    #modele Eric #pb effet albedo pour partitionning vegetation

# rq: pourrait gerer par voxel de surface plutot que globalement (ne pas assumer homogeneite horizontale)


def soil_EV_STICS(Et0, Precip, epsi, previous_state=[0., 0., 0.], leafAlbedo=0.15, U=5., b=0.63):
    """
    
    """
    #""" U: reservoir superficiel = quantite d'eau dans une couche superieure (varie generallement entre 0 et 30 mm
    #    b: coeef empirique sans dimension pour l'evaporation du sol (cf fonction bEV)
    #    meme formulation pour Lebon et al 2003 et STICS"""

    # recup previous state
    previousSES1 = previous_state[0]
    previousSES2 = previous_state[1]
    previousSEP = previous_state[2]

    ##adaptation eq 7.1 p127 STICS book
    k_eq = 0.7  # LAI equivalent pour un k=0.7; pause k car les deux k et LAI sont inconnu
    LAIeq = -log(1 - epsi) / k_eq
    EP = Et0 * exp(-(k_eq - 0.2) * LAIeq)  # potentialSoilEvaporation Eq. 7.1, p127
    # rq: la bonne variable a recuperer a la place de espi serait le taux de couverture ou le transmis vertical pour lequel le correctif 0.2 n'est pas necessaire (Eq. 7.2)

    SES1 = max(0., previousSES1)
    SES2 = max(0., previousSES2)
    SEP = max(0., previousSEP)

    P = Precip
    P1 = Precip

    if (SES1 < U):
        # phase 1
        if (P1 > SES1):
            SES1 = 0.
            SES1 = SES1 + EP
        else:
            SES1 = SES1 - P
            SES1 = SES1 + EP

        if (SES1 > U):
            ER = EP - 0.4 * (SES1 - U)
            SES2 = 0.6 * (SES1 - U)
        else:
            ER = EP
    else:
        # phase 2
        if (P1 > SES2):
            SEP = 0
            P1 = P1 - SES2
            SES1 = U - P1

            if (P1 > U):
                SES1 = EP
            else:
                SES1 = SES1 + EP

            if (SES1 > U):
                ER = EP - 0.4 * (SES1 - U)
                SES2 = 0.6 * (SES1 - U)
            else:
                ER = EP

        else:
            SES21 = ((2 * b * SEP + b ** 2) ** 0.5) - b
            SEP = SEP + EP
            SES20 = ((2 * b * SEP + b ** 2) ** 0.5) - b
            ER = SES20 - SES21

            if (P > 0):
                ESx = 0.8 * P

                if (ESx < ER):
                    ESx = ER + P

                if (ESx > EP):
                    ESx = EP

                ER = ESx
                SES2 = SES2 + ER - P
            else:
                if (ER > EP):
                    ER = EP
                    SES2 = SES2 + ER - P
                else:
                    SES2 = SES2 + ER - P

    state = [SES1, SES2, SEP]
    soilEvaporation = ER
    return soilEvaporation, state
    #combine modele eric et calcul EP avec LAI


# def soil_EV_STICS_old(Et0, Precip, epsi, previous_state=[0., 0., 0.], leafAlbedo=0.15, U=5., b=0.63):
#     """
# 
#     """
#     # """ U: reservoir superficiel = quantite d'eau dans une couche superieure (varie generallement entre 0 et 30 mm
#     #    b: coeef empirique sans dimension pour l'evaporation du sol (cf fonction bEV)
#     #    Et0: Evapotranspiration potenrielle de reference (gazon) en mm
#     #    meme formulation pour Lebon et al 2003 et STICS
#     #    SES1: EP cumule depuis la derniere pluie (previous_state[0])
#     #    SES2: EP cumule depuis passage a phase 2  (previous_state[1])
#     #    SEP : cumul des evap depuis debut phase 2
#     #    ER SoilEvaporation Reel"""
# 
#     ##Riou -> donne des valeur negatives pour forts epsi
#     # coeffTV = epsi/ (1.-leafAlbedo)
#     # EP = (1. - coeffTV) * Et0#potentialSoilEvaporation
# 
#     ##adaptation eq 7.1 p127 STICS book
#     k_eq = 0.7  # LAI equivalent pour un k=0.7; pause k car les deux k et LAI sont inconnu
#     LAIeq = -log(1 - epsi) / k_eq
#     EP = Et0 * exp(-(k_eq - 0.2) * LAIeq)  # potentialSoilEvaporation Eq. 7.1, p127
#     # rq: la bonne variable a recuperer a la place de espi serait le taux de couverture ou le transmis vertical pour lequel le correctif 0.2 n'est pas necessaire (Eq. 7.2)
# 
#     if previous_state[0] + EP < U:  # phase 1
#         SES1 = previous_state[0] + EP
#         SES2 = 0.
#         ER = EP
#         SEP = 0.
#         # mise a jour state!
#     elif previous_state[0] < U and previous_state[0] + EP >= U:  # transition phase1->2
#         fracEP2 = previous_state[0] + EP - U  # precipitation evaporee en phase 2
#         fracEP1 = EP - fracEP2
#         SES1 = previous_state[0] + EP
#         SES2 = fracEP2
#         ER2 = ((2 * b * SES2 + b ** 2) ** 0.5) - b
#         ER = fracEP1 + (fracEP2 / EP) * ER2  # verif
#         SEP = ER2
#     else:  # phase 2
#         SES1 = previous_state[0] + EP
#         SES2 = previous_state[1] + EP
#         ER = ((2 * b * SES2 + b ** 2) ** 0.5) - b - previous_state[
#             2]  # fonction cumul, donc retire valeur de la date precedente
#         SEP = ((2 * b * SES2 + b ** 2) ** 0.5) - b
# 
#     if Precip > 0.:  # comme si pluie vient en fin de journee
#         SES1 = 0.
#         SES2 = 0.
#         SEP = 0.
# 
#     return ER, [SES1, SES2, SEP]  # 0., [SES1, SES2, 0.]#
#     # test par GL
#     # pb de reponse au faiblee pluies: remet tout a zero ; agit fort alors que devrait pas ('effet pompe')


# def soil_EV_STICS_new1(Et0, Precip, epsi, previous_state=[0., 0., 0.], leafAlbedo=0.15, U=5., b=0.63):
#     """
# 
#     """
#     # """ U: reservoir superficiel = quantite d'eau dans une couche superieure (varie generallement entre 0 et 30 mm
#     #    b: coeef empirique sans dimension pour l'evaporation du sol (cf fonction bEV)
#     #    Et0: Evapotranspiration potenrielle de reference (gazon) en mm
#     #    meme formulation pour Lebon et al 2003 et STICS
#     #    SES1: EP cumule depuis la derniere pluie (previous_state[0])
#     #    SES2: EP cumule depuis passage a phase 2  (previous_state[1])
#     #    SEP : cumul des evap depuis debut phase 2
#     #    ER SoilEvaporation Reel"""
# 
#     ##Riou -> donne des valeur negatives pour forts epsi
#     # coeffTV = epsi/ (1.-leafAlbedo)
#     # EP = (1. - coeffTV) * Et0#potentialSoilEvaporation
# 
#     ##adaptation eq 7.1 p127 STICS book
#     k_eq = 0.7  # LAI equivalent pour un k=0.7; pause k car les deux k et LAI sont inconnu
#     LAIeq = -log(1 - epsi) / k_eq
#     EP = Et0 * exp(-(k_eq - 0.2) * LAIeq)  # potentialSoilEvaporation Eq. 7.1, p127
#     # rq: la bonne variable a recuperer a la place de espi serait le taux de couverture ou le transmis vertical pour lequel le correctif 0.2 n'est pas necessaire (Eq. 7.2)
# 
#     if previous_state[0] + EP < U:  # phase 1
#         SES1 = previous_state[0] + EP
#         SES2 = 0.
#         ER = EP
#         SEP = 0.
#         # mise a jour state!
#     elif previous_state[0] < U and previous_state[0] + EP >= U:  # transition phase1->2
#         fracEP2 = previous_state[0] + EP - U  # precipitation evaporee en phase 2
#         fracEP1 = EP - fracEP2
#         SES1 = previous_state[0] + EP
#         SES2 = fracEP2
#         ER2 = ((2 * b * SES2 + b ** 2) ** 0.5) - b
#         ER = fracEP1 + (fracEP2 / EP) * ER2  # verif
#         SEP = ER2
#     else:  # phase 2
#         SES1 = previous_state[0] + EP
#         SES2 = previous_state[1] + EP
#         ER = ((2 * b * SES2 + b ** 2) ** 0.5) - b - previous_state[
#             2]  # fonction cumul, donc retire valeur de la date precedente
#         SEP = ((2 * b * SES2 + b ** 2) ** 0.5) - b
# 
#     # MAJ memoire si pluie - comme si pluie vient en fin de journee, apres calcul ER
#     seuil = U  # 0.7*U
#     if Precip > 0. and Precip >= seuil and SES1 > U:  # si en phase 2 et forte precipitation
#         SES1 = 0.
#         SES2 = 0.
#         SEP = 0.
#     elif Precip > 0. and Precip < seuil and SES1 > U:  # si en phase 2 et faible precipitation
#         # remet en debut de phase 2
#         SES1 = SES1 - SES2
#         SES2 = 0.
#         SEP = 0.
#     elif Precip > 0. and SES1 <= U:  # si en phase 1
#         SES1 = 0.
#         SES2 = 0.
#         SEP = 0.
# 
#     return ER, [SES1, SES2, SEP]  # 0., [SES1, SES2, 0.]#
#     # a tester pour limiter effet peites pluie - effet trop fort! + seuil arbitraire avec tres fort impact


#def evap_frac_zone(dz = [0., 0.3], ZESX=0.3):
#    """ distribution en profondeur du prelevement d'eau evapore sous un voxel de surface
#        integrale de y = -ZESX*ZESX*Z/2 + ZESX -> relation lineaire de la figure 7.4b STICS manual (p130)
#        correspond a equation 7.6 pour CFES=1 (lineaire)"""
#    if dz[0]>=0. and dz[0]<=ZESX: #1ere borne OK
#        if dz[1]>0. and dz[1]<=ZESX: #2e borne OK
#            frac_interval = (-(1./ZESX**2)*(dz[1]**2)+(2./ZESX)*dz[1]) - (-(1./ZESX**2)*(dz[0]**2)+(2./ZESX)*dz[0])
#        elif dz[1]>0. and dz[1]>ZESX: #2e superieure-> remplace par profondeur max ZESX
#            frac_interval = (-(1./ZESX**2)*(ZESX**2)+(2./ZESX)*ZESX) - (-(1./ZESX**2)*(dz[0]**2)+(2./ZESX)*dz[0])
#        else:
#            frac_interval = 0.
#    else:
#        frac_interval = 0.
#
#    return frac_interval

#evap_frac_zone(dz = [0., 0.2], ZESX=0.3)


#def check_soil_evap(evapo_tot, m_frac_evap, tsw_t, m_QH20min, dxyz, ZESX = 0.3):
#    """ verifie que evaporation du sol n'entraine pas de tsw sous la tsw mini d'un sol sec (air) ; sinon ajuste distribution en profondeur(report sur la couche inferieure), puis evapo_tot"""
#    m_evap = deepcopy(m_frac_evap)
#
#    #recherche des id couches sur lesquelles evaporation joue
#    nz, cumz = 0, []
#    for i in range(len(dxyz[2])):
#        cumz.append(dxyz[2][i])
#        nz = nz+1
#        if sum(cumz)>=ZESX:
#            break 
#
#    for x in range(len(dxyz[0])):
#        for y in range(len(dxyz[1])):
#            for z in range(nz-1):#report par colone de sol
#                #res = tsw_t[z][x][y] - m_evap[z][x][y]#m_frac_evap[z][x][y] #- report
#                res = tsw_t[z][x][y] - m_QH20min[z][x][y] - m_evap[z][x][y]#m_frac_evap[z][x][y] #- report
#                if res<=0.:
#                    #m_evap[z+1][x][y] = m_evap[z+1][x][y] + m_evap[z][x][y] - tsw_t[z][x][y]
#                    #m_evap[z][x][y] = tsw_t[z][x][y]
#                    m_evap[z+1][x][y] = m_evap[z+1][x][y]-m_QH20min[z+1][x][y] + m_evap[z][x][y]-m_QH20min[z][x][y] - tsw_t[z][x][y]#??
#                    m_evap[z][x][y] = tsw_t[z][x][y]- m_QH20min[z][x][y]#??
#
#            #derniere couche affectee
#            z = nz-1
#            res = tsw_t[z][x][y] - m_QH20min[z][x][y] - m_evap[z][x][y]#m_frac_evap[z][x][y] #- report
#            if res<=0.:
#                m_evap[z][x][y] = tsw_t[z][x][y] - m_QH20min[z][x][y]
#
#    return sum3(m_evap), m_evap

#test = deepcopy(S.tsw_t)
#test[0][0][0]=0.
#test[1][0][0]=0.
#test[2][0][0]=0.
#test[3][0][0]=0.
#test[4][0][0]=0.#002
#test[5][0][0]=0.#002
#test[6][0][0]=0.#002
#test[7][0][0]=0.002
#evapo_tot,  m_evap = check_soil_evap(sum3(m_frac_evap), m_frac_evap, test, S.dxyz, ZESX = 0.3)


#def distrib_evapNC(m_soil_vox, m_soil_vol, map_Evap0, ZESX=0.3):
#    """ map_Evap0 = liste des Evap des voxel de la couche superieure (matrice x, y) """
#    m = deepcopy(m_soil_vol)
#    if len(m)>1:
#        m.fill(0.)
#        for z in range(len(m)-1):
#            dz = [abs(m_soil_vox[z][0][0][2]), abs(m_soil_vox[z+1][0][0][2])]
#            frac = evap_frac_zone(dz , ZESX)
#            for x in range(len(m[z])):
#                for y in range(len(m[z][x])):
#                    m[z][x][y] = frac*map_Evap0[x][y]
#    else:
#        m.fill(1.)
#
#    return m

#map_Evap0 = array([[1.5]]) ## faire une fonction qui le genere initialement
#m_frac_evap = distrib_evapNC(m_soil_vox, m_soil_vol, map_Evap0, ZESX=0.3)





########## diverses fonction - soil nitrogen balance



## soil N supply


def Convective_Nflux(SN, ls_mWaterUptakePlt):
    """
        
    """
    #""" Eq. 8_34 p 160 - potential passive NO3 uptake with water flow"""
    ls_conv = []
    convtot = SN.m_1*0.
    for i in range(len(ls_mWaterUptakePlt)):
        flux_c = ls_mWaterUptakePlt[i] * SN.ConcNO3()
        ls_conv.append(flux_c)
        convtot = convtot+flux_c

    return convtot, ls_conv
    #renvoyer liste de fraction?

def Diffusive_Nflux(SN, parSN, ls_lrac):
    """
        
    """
    #""" Eq. 8.35 p 161 """
    #ls_lrac = ls_ roots (en m) -> convert en cm (*100)

    ls_diff = []
    difftot = SN.m_1*0.

    FH = (SN.tsw_t - SN.m_QH20wp) / (SN.m_QH20fc - SN.m_QH20wp)
    filtre = FH>0
    filtre = filtre * 1. #remplace valeur negatives par zero
    DIFE = parSN['DIFNg'] * FH * filtre
    coeff = 4*pi**0.5
    for i in range(len(ls_lrac)):
        draci = (ls_lrac[i]*100.) / (SN.m_soil_vol*1000000.) #root density (cm.cm-3)
        flux_d = coeff * DIFE * (SN.m_NO3 + SN.m_NH4) *sqrt(draci)
        ls_diff.append(flux_d)
        difftot = difftot + flux_d

    return difftot, ls_diff
    #a priori ls_diff pas utilise


def Nsoil_supply(SN, parSN, ls_lrac, ls_mWaterUptakePlt):
    """
        
    """
    #""" Eq. 8.33 p 160 """
    convtot, ls_conv = Convective_Nflux(SN, ls_mWaterUptakePlt)
    difftot, ls_diff = Diffusive_Nflux(SN, parSN, ls_lrac)
    ls_suptot = []
    for i in range(len(ls_conv)):
        ls_suptot.append(ls_conv[i] + ls_diff[i])

    return convtot+difftot, ls_suptot
    #a priori ls_suptot pas utilise

